source/program/main.cpp

```diff
@@ -146,55 +146,141 @@ int multiworld_init(lua_State* L) {
-            }
+                outputSuccess = true;
+            } 
+            sendBuffer->insert(sendBuffer->begin(), luaResult, luaResult + resultSize);
         } else {
-            response = snprintf(output_start, output_buffer_size, "error parsing buffer: %d", loadResult);
+            std::string errorMessage = "error parsing buffer: " + std::to_string(loadResult);
+            const char* errorAsCString = errorMessage.c_str();
+            resultSize = strlen(errorAsCString);
+            sendBuffer->insert(sendBuffer->begin(), errorAsCString, errorAsCString + resultSize);
         }
-        buffer[0] = output_success;        
-        buffer[1] = response & 0xff;
-        buffer[2] = (response >> 8)  & 0xff;
-        buffer[3] = (response >> 16) & 0xff;
-        return response + 4;
+        sendBuffer->insert(sendBuffer->begin(), PACKET_REMOTE_LUA_EXEC);
+        sendBuffer->insert(sendBuffer->begin() + 1, outputSuccess);
+        sendBuffer->insert(sendBuffer->begin() + 2, resultSize & 0xff);
+        sendBuffer->insert(sendBuffer->begin() + 3, (resultSize >> 8)  & 0xff);
+        sendBuffer->insert(sendBuffer->begin() + 4, (resultSize >> 16)  & 0xff);
+        return sendBuffer;
     });
 
     // Register calling update again
     multiworld_schedule_update(L);
     return 0;
 }
 
+std::vector<u8>* create_packet_from_lua_string(lua_State* L) {
+    size_t resultSize = 0;          // length of the lua string response (without \0)
+    std::vector<u8>* sendBuffer;  // sendBuffer to store the result. this pointer is returned 
+    
+    const char* luaResult = lua_tolstring(L, 1, &resultSize);
+
+    sendBuffer = new std::vector<u8>();
+    if (sendBuffer == NULL) return NULL;
+
+    uint32_t resultAs32Bit = resultSize;
+    u8* chars = reinterpret_cast<u8*>(&resultAs32Bit);
+    // If you move the following line by one line down the compiler throws a warning?!?
+    sendBuffer->insert(sendBuffer->begin(), luaResult, luaResult + resultAs32Bit);
+    sendBuffer->insert(sendBuffer->begin(), chars, chars + sizeof(uint32_t));
+
+    return sendBuffer;
+}
+
+/* This function gets called by the lua to sent message to the client */
+int gamelog_send(lua_State* L) {
+    if (RemoteApi::clientSubs.logging) {
+        RemoteApi::SendMessage([=]() -> std::vector<u8>* {
+            std::vector<u8>* sendBuffer = create_packet_from_lua_string(L); 
+            if (sendBuffer == NULL) return NULL;
+            sendBuffer->insert(sendBuffer->begin(), (u8)PACKET_LOG_MESSAGE);
+            return sendBuffer;
+        });
+    }
+    return 1;
+}
+
+/* This function gets called by the lua to sent message to the client */
+int inventory_send(lua_State* L) {
+    if (RemoteApi::clientSubs.multiWorld) {
+        RemoteApi::SendMessage([=]() -> std::vector<u8>* {
+            std::vector<u8>* sendBuffer = create_packet_from_lua_string(L); 
+            if (sendBuffer == NULL) return NULL;

```

Comment:
With `unique_ptr`, this could just be:
```c++
        RemoteApi::SendMessage([](lua_State* L) -> PacketBuffer {
            PacketBuffer sendBuffer = create_packet_from_lua_string(L); 
            if (sendBuffer) {
                sendBuffer->insert(sendBuffer->begin(), (u8)PACKET_COLLECTED_INDICES);
            }
            return sendBuffer;
        });
```